### PR TITLE
First aproach of rust drivers for AK09915

### PR DESCRIPTION
**Pull request by RaulTrombin**
This driver has the capability to perform the following functions:

- perform a software reset
- switch between modes
- check if data is ready for reading
- read data from the magnetometer

Since this driver was designed to be agnostic, it can be tested in the "examples" folder for Raspberry Pi using the "rpi-test.sh" script. You can modify the DNS and PORT settings in this script as needed.

image

### CODE REVIEW

File: examples/raspberry-pi/rpi-test.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+# Define variables for DNS and port
+DNS=
```
Comment (patrickelectric) [inline examples/raspberry-pi/rpi-test.sh:4]: this should be `ADDRESS` or something more clear, maybe `IP`:

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    // When Data Ready, proceed to the next step.
+    // (4)Read measurement data (HXL to HZH)
+    // 9.4.4.2. Self-test Judgment
+    // When measurement data read by the above sequence is in the range of following table, AK09915 is working
+    // normally.
+    // HX[15:0] bits HY[15:0] bits HZ[15:0] bits
+    // Criteria -200 ≤HX≤ +200 -200 ≤HY≤ +200 -800 ≤HZ≤ -200
+
+    pub fn self_test(&mut self) -> Result<(), E> {
+        self.set_mode(Mode::SelfTest)?;
```
[the remark this answers is not in the document]
Comment (RaulTrombin) [reply src/lib.rs:127]: Yes, you are right, but it already does inside set_mode(),
it set power down, wait 100uS and change to new mode.

File: examples/raspberry-pi/rpi-test.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+PORT=
```
Comment (patrickelectric) [inline examples/raspberry-pi/rpi-test.sh:5]: Use default values:
```suggestion
PORT="${PORT:-22}"
```

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: examples/raspberry-pi/rpi-test.sh (hunk 1 above)
Comment (joaoantoniocardoso) [reply examples/raspberry-pi/rpi-test.sh:4]: > this should be `ADDRESS` or something more clear, maybe `IP`:

if only ipv4 and ipv6 are intended, use _ip_. If both DNSs and IPs, _address_. If any web stuff, _url_,

**Comment by patrickelectric** [comment]
please rebase the code over master

File: examples/raspberry-pi/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    //tests if single mode are being
+
+    println!("Test 5 single measurement, and 5 without set single measurement(test data ready"); 
+    for n in 1..=5 {
+        sensor.set_mode(Mode::SingleMeasurement)?;
+        if (sensor.is_data_ready()?){
+            let (x, y, z) = sensor.read_mag()?;
+            println!("Magnetometer: x={}, y={}, z={}", x, y, z);
+        }
+        }
```
[the remark this answers is not in the document]
Comment (RaulTrombin) [reply examples/raspberry-pi/src/main.rs:31]: image

File: examples/raspberry-pi/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+use ak09915::Ak09915;
+use ak09915::Mode;
+use linux_embedded_hal::I2cdev;
+
+
+fn main() -> Result<(), linux_embedded_hal::i2cdev::linux::LinuxI2CError> {
+    let dev = I2cdev::new("/dev/i2c-1")?;
```
[the remark this answers is not in the document]
Comment (RaulTrombin) [reply examples/raspberry-pi/src/main.rs]: image

File: examples/raspberry-pi/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "raspberry-pi"
```
[the remark this answers is not in the document]
Comment (RaulTrombin) [reply examples/raspberry-pi/Cargo.toml:2]: It can be done as examples/raspberrypi.rs, but need add some .cargo settings to library folder(such linkers and other settings). 
Should I untrack this project inside ? (I'm using to debug/test, since the driver is agnostic)

File: examples/raspberry-pi/src/main.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (RaulTrombin) [reply examples/raspberry-pi/src/main.rs]: image
done!

File: examples/raspberry-pi/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (RaulTrombin) [reply examples/raspberry-pi/Cargo.toml:2]: It's possible, but will need add unnecessary settings to main library, It's was being used to test/debug the lib, since the drivers are agnostic made. 

Should untrack the project from library?

**Review by RaulTrombin — COMMENTED** [review]
All suggestions where applyed,
Need to verify if keep the project example inside lib.

File: examples/raspberry-pi/Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (patrickelectric) [reply examples/raspberry-pi/Cargo.toml]: if we move this code to an example we can remove mostly of the files here.

File: examples/raspberry-pi/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,48 @@
+use ak09915::Ak09915;
+use ak09915::Mode;
+use linux_embedded_hal::I2cdev;
+use clap::Parser;
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Receive the i2c device as parameter
+    #[arg(short, long, default_value_t = ("/dev/i2c-1").to_string())]
```
Comment (patrickelectric) [inline examples/raspberry-pi/src/main.rs]: use `default_value`

File: src/lib.rs [hunk 8]
```diff
@@ -1,14 +1,186 @@
...
+    I2C: Write<Error = E> + WriteRead<Error = E>,
+{
+    pub fn new(i2c: I2C) -> Self {
+        Self {
+            i2c,
+            address: AK09915_ADDRESS,
+        }
+    }
+
+    fn write_register(&mut self, register: u8, value: u8) -> Result<(), E> {
```
Comment (patrickelectric) [inline src/lib.rs:68]: This should use `Register as argument`

File: src/lib.rs [hunk 9]
```diff
@@ -1,14 +1,186 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use embedded_hal::blocking::i2c::{Write, WriteRead};
+
+const AK09915_ADDRESS: u8 = 0x0C;
+
+// Register Addresses
+#[allow(dead_code)]
+enum Register {
```
Comment (patrickelectric) [inline src/lib.rs:7]: missing `#[repr(u16)]`

File: src/lib.rs [hunk 10]
```diff
@@ -1,14 +1,186 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use embedded_hal::blocking::i2c::{Write, WriteRead};
+
+const AK09915_ADDRESS: u8 = 0x0C;
+
+// Register Addresses
+#[allow(dead_code)]
```
Comment (patrickelectric) [inline src/lib.rs:6]: Why is this necessary ?

File: src/lib.rs [hunk 11]
```diff
@@ -1,14 +1,186 @@
...
+    }
+
+    pub fn reset(&mut self) -> Result<(), E> {
+        // Soft reset device
+        self.write_register(u8::from(Register::CNTL3), 0x01)?;
+        Ok(())
+    }
+
+    pub fn set_mode(&mut self, mode: Mode) -> Result<(), E> {
+        self.write_register(u8::from(Register::CNTL2), u8::from(Mode::PowerDown))?;
```
Comment (patrickelectric) [inline src/lib.rs:93]: you can use just `as u8` or `into()`, that are less verbose.

File: src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+        self.is_data_ready()?;
+
+        let (hx, hy, hz) = self.read_mag()?;
+
+        // Self-test judgment
+        if (hx >= -200 && hx <= 200) && (hy >= -200 && hy <= 200) && (hz >= -800 && hz <= -200) {
+            println!("Self-test passed \nMagnetometer: x={}, y={}, z={}", hx, hy, hz);
+        } else {
+            println!("Self-test failed");
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply src/lib.rs:129]: Why is this marked as solved ?

File: src/lib.rs [hunk 13]
```diff
@@ -1,14 +1,186 @@
...
+        let (hx, hy, hz) = self.read_mag()?;
+
+        // Self-test judgment
+        if (hx >= -200 && hx <= 200) && (hy >= -200 && hy <= 200) && (hz >= -800 && hz <= -200) {
+            println!(
+                "Self-test passed \nMagnetometer: x={}, y={}, z={}",
+                hx, hy, hz
+            );
+            return Ok(true);
+        } else {
```
Comment (patrickelectric) [inline src/lib.rs:128]: unnecessary else

**Review by patrickelectric — CHANGES_REQUESTED** [review]
The commits should also be squashed

File: src/lib.rs (hunk 10 above)
Comment (RaulTrombin) [reply src/lib.rs:6]: just to supress unnused registers warning

File: examples/raspberry-pi/src/main.rs (hunk 7 above)
Comment (RaulTrombin) [reply examples/raspberry-pi/src/main.rs]: ok, fixed.

File: src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (joaoantoniocardoso) [reply src/lib.rs:129]: Better to be a Result<()> because `set_mode` can fail. Then, instead of the `false`, it can express what went wrong with a message/context.

File: src/lib.rs [hunk 14]
```diff
@@ -1,14 +1,186 @@
...
+        // Create a mock I2C device and queue the expected transactions
+        let i2c_mock = I2cMock::new(&expected_trans);
+
+        // Create an AK09915 instance using the mock I2C device
+        let mut sensor = Ak09915::new(i2c_mock);
+
+        // Set the mode to continuous measurement mode 1
+        // sensor.set_mode(Mode::Cont1Hz).unwrap();
+
+        let (x, y, z) = sensor.read_mag().unwrap();
```
Comment (joaoantoniocardoso) [inline src/lib.rs:179]: an `expect` instead of a `unwrap` would make an easier-to-debug test.

File: src/lib.rs [hunk 15]
```diff
@@ -1,14 +1,186 @@
...
+        ];
+
+        // Create a mock I2C device and queue the expected transactions
+        let i2c_mock = I2cMock::new(&expected_trans);
+
+        // Create an AK09915 instance using the mock I2C device
+        let mut sensor = Ak09915::new(i2c_mock);
+
+        // Set the mode to continuous measurement mode 1
+        // sensor.set_mode(Mode::Cont1Hz).unwrap();
```
Comment (joaoantoniocardoso) [inline src/lib.rs:177]: should this be commented out?

File: src/lib.rs [hunk 16]
```diff
@@ -1,14 +1,186 @@
...
+
+    pub fn init(&mut self) -> Result<(), E> {
+        // Soft reset device and put on continuous measurement
+        self.reset()?;
+        self.set_mode(Mode::PowerDown)?;
+        Ok(())
+    }
+
+    pub fn reset(&mut self) -> Result<(), E> {
+        // Soft reset device
```
Comment (joaoantoniocardoso) [inline src/lib.rs:87]: Hint: use [doc comments syntax](https://doc.rust-lang.org/rust-by-example/meta/doc.html) here and everywhere else :)
```suggestion

    /// Soft reset device
    pub fn reset(&mut self) -> Result<(), E> {
```

Comment (joaoantoniocardoso) [reply src/lib.rs:87]: actually, might be better to do it in another PR, if/when we decide to prepare this to be a published crate to crates.io, then we just follow this: https://doc.rust-lang.org/book/ch14-02-publishing-to-crates-io.html#publishing-a-crate-to-cratesio

File: examples/linux-i2c.rs [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+use clap::Parser;
+use linux_embedded_hal::I2cdev;
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Receive the i2c device as parameter
+    #[arg(short, long, default_value_t = ("/dev/i2c-1").to_string())]
+    device: String,
+}
```
Comment (joaoantoniocardoso) [inline examples/linux-i2c.rs:12]: Cool, are those commands automatically generated from the info contained in the cargo.toml file? Also, instead of author and about commands, wouldn't be better a `help` one?

Comment (patrickelectric) [reply examples/linux-i2c.rs:12]: Helper is generated by default. Our projects were created before clap had support to structs directly, before that was another crate.

File: src/lib.rs (hunk 8 above)
Comment (patrickelectric) [inline src/lib.rs:68]: This should use Register as argument

File: src/lib.rs (hunk 9 above)
Comment (RaulTrombin) [reply src/lib.rs:7]: fixed

File: src/lib.rs (hunk 10 above)
Comment (RaulTrombin) [reply src/lib.rs:6]: it's fine?

File: src/lib.rs (hunk 11 above)
Comment (RaulTrombin) [reply src/lib.rs:93]: fixed

File: src/lib.rs (hunk 13 above)
Comment (RaulTrombin) [reply src/lib.rs:128]: removed

File: src/lib.rs (hunk 15 above)
Comment (RaulTrombin) [reply src/lib.rs:177]: I'm having issues with the i2c mock for especial functions, 
set_mode change 2 registers and have a delay inside, 
something went wrong and commented until fix it's aproach

File: src/lib.rs (hunk 8 above)
Comment (RaulTrombin) [reply src/lib.rs:68]: fixed

File: src/lib.rs [hunk 18]
```diff
@@ -1,14 +1,187 @@
...
+    fn write_register(&mut self, register: Register, value: u8) -> Result<(), E> {
+        self.i2c.write(self.address, &[register.into(), value])
+    }
+
+    fn read_register(&mut self, register: Register) -> Result<u8, E> {
+        let mut buffer = [0u8];
+        self.i2c
+            .write_read(self.address, &[register.into()], &mut buffer)
+            .and(Ok(buffer[0]))
+    }
```
Comment (patrickelectric) [inline src/lib.rs:79]: this functions should be public as well to make the project more versatile

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 19]
```diff
@@ -1,2 +1,2 @@
 /target
-/Cargo.lock
+/Cargo.lock
```
Comment (patrickelectric) [inline .gitignore:2]: this change and commit can be removed.

File: Cargo.toml [hunk 20]
```diff
@@ -4,3 +4,4 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
+embedded-hal = "0.2.5"
```
Comment (patrickelectric) [inline Cargo.toml:7]: We usually use prefix in our commits, such as: `Cargo: Add embedded-hal`

File: Cargo.toml [hunk 21]
```diff
@@ -5,3 +5,8 @@ edition = "2021"
 
 [dependencies]
 embedded-hal = "0.2.5"
+
+[dev-dependencies]
+embedded-hal-mock = "0.9.0"
+clap = { version = "4.2.2", features = ["derive"] }
+linux-embedded-hal = "0.3.2"
```
Comment (patrickelectric) [inline Cargo.toml:12]: please sort the library names

File: src/lib.rs [hunk 22]
```diff
@@ -109,19 +109,20 @@ where
...
+    //      - Polling DRDY bit of ST1 register.
+    //      - Monitoring DRDY pin.
+    //      When Data Ready, proceed to the next step.
+    //   4. Read measurement data (HXL to HZH).
+    // 9.4.4.2. Self-test Judgment:
+    //   If measurement data read by the above sequence is within the following ranges,
+    //   AK09915 is working normally:
+    //     - HX[15:0] bits: -200 ≤ HX ≤ +200
+    //     - HY[15:0] bits: -200 ≤ HY ≤ +200
+    //     - HZ[15:0] bits: -800 ≤ HZ ≤ -200
```
Comment (patrickelectric) [inline src/lib.rs:125]: This should be squashed with: e61d15c792036f294828edb3cc80284f4e537842
Do not create commits for things that only exist in the PR.

**Review by patrickelectric — CHANGES_REQUESTED** [review]
Please squash the commits on the correct commits

File: src/lib.rs [hunk 23]
```diff
@@ -1,14 +1,234 @@
...
+pub struct Ak09915<I2C> {
+    pub i2c: I2C,
+    pub address: u8,
+}
+
+impl<I2C, E> Ak09915<I2C>
+where
+    I2C: Write<Error = E> + WriteRead<Error = E>,
+{
+    // type Error = ();
```
Comment (patrickelectric) [inline src/lib.rs:70]: lost comment

File: src/lib.rs [hunk 24]
```diff
@@ -1,14 +1,234 @@
...
+    }
+    // 9.4.3.2. Normal Read Sequence:
+    //   1. Check Data Ready or not by any of the following method:
+    //      |ST2| -> | 0 0 0 0 HOFL INV 0 0 |
+    pub fn check_overflow(&mut self) -> Result<(), Error<E>> {
+        let status = self.read_register(Register::ST2)?;
+        if (status & 0x04) != 0 {
+            return Err(Error::SensorOverflow); // Data not ready after retries
+        }
+        Ok(()) // Normal
```
Comment (patrickelectric) [inline src/lib.rs:163]: unnecessary comment

File: src/lib.rs [hunk 25]
```diff
@@ -1,14 +1,234 @@
...
+    Cont50Hz = 0x06,
+    Cont100Hz = 0x08,
+    Cont200Hz = 0x0A,
+    Cont1Hz = 0x0C,
+    SelfTest = 0x10,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Error<E> {
+    I2C(E), // I²C bus error
```
Comment (patrickelectric) [inline src/lib.rs:44]: unnecessary comment

**Review by patrickelectric — CHANGES_REQUESTED** [review]

**Review by patrickelectric — APPROVED** [review]
